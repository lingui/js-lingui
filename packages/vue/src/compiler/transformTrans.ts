import {
  createSimpleExpression,
  type ElementNode,
  NodeTypes,
  ConstantTypes,
  ElementTypes,
  SourceLocation,
  type TemplateNode,
} from "@vue/compiler-core"
import {
  createMacroTransVueContext,
  tokenizeTrans,
  getTextProp,
} from "../common/Trans"
import {
  ICUMessageFormat,
  createMessageDescriptor,
  JsMacroName,
} from "@lingui/babel-plugin-lingui-macro/ast"
import generate from "@babel/generator"
import { ObjectExpression } from "@babel/types"

function wrapInTemplateSlotNode(
  index: number,
  child: ElementNode
): TemplateNode {
  const loc = child.loc
  return {
    type: NodeTypes.ELEMENT,
    ns: 0,
    tag: "template",
    tagType: ElementTypes.TEMPLATE,
    props: [
      {
        type: NodeTypes.DIRECTIVE,
        name: "slot",
        exp: createSimpleExpression("{children}", false, loc, 0),
        arg: createSimpleExpression(String(index), false, loc, 3),
        modifiers: [],
        loc,
      },
    ],
    isSelfClosing: false,
    children: [child],
    codegenNode: undefined,
    loc,
  }
}

export function createInnerSlotNode(sourceChild: ElementNode): ElementNode {
  if (sourceChild.isSelfClosing) return sourceChild
  const loc = sourceChild.loc
  // no need for a deep copy
  return {
    ...sourceChild,
    children: [
      {
        type: NodeTypes.ELEMENT,
        ns: 0,
        tag: "component",
        tagType: ElementTypes.COMPONENT,
        props: [
          {
            type: NodeTypes.DIRECTIVE,
            name: "bind",
            exp: createSimpleExpression("children", false, loc, 0),
            arg: createSimpleExpression("is", true, loc, 3),
            modifiers: [],
            loc,
          },
        ],
        isSelfClosing: false,
        children: [],
        loc,
        codegenNode: undefined,
      },
    ],
  }
}

function transformElement(
  node: ElementNode,
  descriptor: ObjectExpression,
  loc: SourceLocation,
  elements?: Record<string, ElementNode> | undefined
) {
  const simpleExp = createSimpleExpression(
    generate(descriptor).code,
    false,
    loc,
    ConstantTypes.NOT_CONSTANT
  )

  node.props = [
    {
      type: NodeTypes.DIRECTIVE,
      name: "bind",
      exp: simpleExp,
      arg: undefined,
      modifiers: [],
      loc,
    },
  ]

  node.isSelfClosing = true

  if (elements) {
    node.children = Object.keys(elements).map((key) => {
      return wrapInTemplateSlotNode(
        Number(key),
        createInnerSlotNode(elements[key]! as any)
      )
    })
  } else {
    node.children = []
  }

  node.tagType = ElementTypes.COMPONENT
}

// from SFC to Babel
function convertLoc(
  loc: SourceLocation
): Parameters<typeof createMessageDescriptor>[1] {
  return {
    start: { ...loc.start, index: loc.start.offset },
    end: { ...loc.end, index: loc.end.offset },
    filename: loc.source,
    identifierName: null,
  }
}

export function transformTrans(
  node: ElementNode,
  options: {
    stripNonEssentialProps?: boolean
  }
) {
  const loc = node.loc
  const tokens = tokenizeTrans(
    node,
    createMacroTransVueContext((identifier, macro) => {
      if (macro === JsMacroName.t) {
        return identifier.name === "vt"
      }

      return identifier.name === macro
    }, !!options.stripNonEssentialProps)
  )

  const messageFormat = new ICUMessageFormat()
  const { message, values, elements } = messageFormat.fromTokens(tokens)

  const id = getTextProp(node, "id")
  const context = getTextProp(node, "context")
  const comment = getTextProp(node, "comment")

  const descriptor = createMessageDescriptor(
    { message, values },
    convertLoc(loc),
    false,
    {
      ...(id
        ? {
            id: {
              text: id,
            },
          }
        : {}),
      ...(context
        ? {
            context: {
              text: context,
            },
          }
        : {}),
      ...(comment
        ? {
            comment: {
              text: comment,
            },
          }
        : {}),
    }
  )

  transformElement(node, descriptor, loc, elements)
}
