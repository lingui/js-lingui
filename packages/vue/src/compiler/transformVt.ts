import {
  createSimpleExpression,
  SourceLocation,
  ConstantTypes,
  ExpressionNode,
} from "@vue/compiler-core"
import * as t from "@babel/types"
import generate from "@babel/generator"
import {
  JsMacroName,
  createMessageDescriptorFromTokens,
  isLinguiIdentifier,
  tokenizeTemplateLiteral,
  createMacroJsContext,
  isChoiceMethod,
  tokenizeChoiceComponent,
  processDescriptor,
  MacroJsContext,
} from "@lingui/babel-plugin-lingui-macro/ast"

const RUNTIME_VT_SYMBOL = "vt._"

function createLinguiSimpleExpression(
  messageDescriptor: t.ObjectExpression,
  oldLoc: SourceLocation
) {
  const ast = t.callExpression(t.identifier(RUNTIME_VT_SYMBOL), [
    messageDescriptor,
  ])

  return createSimpleExpression(
    generate(ast).code,
    false,
    oldLoc,
    ConstantTypes.NOT_CONSTANT
  )
}

function createVueMacroContext(options: { stripNonEssentialProps?: boolean }) {
  return createMacroJsContext((identifier, macro) => {
    if (macro === JsMacroName.t) {
      return identifier.name === "vt"
    }

    return identifier.name === macro
  }, !!options?.stripNonEssentialProps)
}

export function tokenizeAsChoiceComponentOrUndefined(
  node: t.Node,
  ctx: MacroJsContext
) {
  // plural / select / selectOrdinal
  if (t.isCallExpression(node) && isChoiceMethod(node, ctx)) {
    return [tokenizeChoiceComponent(node, isChoiceMethod(node, ctx)!, ctx)]
  }

  return
}

export function tokenizeAsLinguiTemplateLiteralOrUndefined(
  node: t.Node,
  ctx: MacroJsContext
) {
  // t`Hello!`
  if (
    t.isTaggedTemplateExpression(node) &&
    isLinguiIdentifier(node.tag, JsMacroName.t, ctx)
  ) {
    return tokenizeTemplateLiteral(node, ctx)
  }
  return
}

export function transformVt(
  vueNode: ExpressionNode,
  options: {
    stripNonEssentialProps?: boolean
  }
) {
  const ctx = createVueMacroContext(options)
  if (!vueNode.ast) {
    return vueNode
  }

  const node = vueNode.ast

  // plural / select / selectOrdinal
  const choiceCmpTokens = tokenizeAsChoiceComponentOrUndefined(node, ctx)

  if (choiceCmpTokens) {
    const messageDescriptor = createMessageDescriptorFromTokens(
      choiceCmpTokens,
      node.loc!,
      ctx.stripNonEssentialProps
    )

    return createLinguiSimpleExpression(messageDescriptor, vueNode.loc!)
  }

  //  t(...)
  if (
    t.isCallExpression(node) &&
    isLinguiIdentifier(node.callee, JsMacroName.t, ctx)
  ) {
    const messageDescriptor = processDescriptor(
      node.arguments[0] as t.ObjectExpression,
      ctx
    )

    return createLinguiSimpleExpression(messageDescriptor, vueNode.loc!)
  }

  // t`Hello!`
  const tplLiteralTokens = tokenizeAsLinguiTemplateLiteralOrUndefined(node, ctx)
  if (tplLiteralTokens) {
    const messageDescriptor = createMessageDescriptorFromTokens(
      tplLiteralTokens,
      node.loc!,
      ctx.stripNonEssentialProps
    )

    return createLinguiSimpleExpression(messageDescriptor, vueNode.loc!)
  }

  return vueNode
}
