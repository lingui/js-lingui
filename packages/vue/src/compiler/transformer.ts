import { type NodeTransform, NodeTypes } from "@vue/compiler-core"

import {
  isTrans,
  isElementNode,
  isDirectiveNode,
  isInterpolationNode,
} from "../common/predicates"
import { transformVt } from "./transformVt"
import { transformTrans } from "./transformTrans"

export type LinguiTransformerOptions = {
  stripNonEssentialProps?: boolean
}

/**
 * Here is an entry point transformer.
 * We need our template transformer to operate on user authored code
 * before any other Vue transformer process it.
 *
 * So this transformer unshifts a real transformer to the transformers array.
 *
 */
export const transformer =
  (options: LinguiTransformerOptions = {}): NodeTransform =>
  (node, context) => {
    if (
      node.type === NodeTypes.ROOT &&
      !context.nodeTransforms.find(
        (t) => (t as any).__name === "linguiTransform"
      )
    ) {
      context.nodeTransforms.unshift(templateTransformer(options))
    }
  }

/**
 * Actual transformer expanding macro calls.
 */
const templateTransformer = (
  options: LinguiTransformerOptions
): NodeTransform => {
  const transformer: NodeTransform = (node) => {
    if (isElementNode(node)) {
      if (isTrans(node)) {
        transformTrans(node, options)
      }

      for (const prop of node.props) {
        if (isDirectiveNode(prop) && prop.exp) {
          prop.exp = transformVt(prop.exp, options)
        }
      }
    }

    if (isInterpolationNode(node)) {
      node.content = transformVt(node.content, options)
    }
  }

  ;(transformer as any).__name = "linguiTransform"
  return transformer
}
