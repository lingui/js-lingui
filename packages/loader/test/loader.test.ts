import path from "path"
import fs from "node:fs/promises"
import { build, watch } from "./compiler"
import { mkdtempSync } from "fs"
import os from "os"

const skipOnWindows = os.platform() === "win32" ? describe.skip : describe

skipOnWindows("lingui-loader", () => {
  it("should compile catalog in po format", async () => {
    const built = await build(path.join(__dirname, "po-format/entrypoint.js"))

    const data = await built.loadBundle()
    expect(built.stats.errors).toEqual([])
    expect(built.stats.warnings).toEqual([])

    expect((await data.load()).messages).toMatchSnapshot()
  })

  it("should compile catalog in json format", async () => {
    const built = await build(
      path.join(__dirname, "./json-format/entrypoint.js")
    )

    expect(built.stats.errors).toEqual([])
    expect(built.stats.warnings).toEqual([])

    const data = await built.loadBundle()
    expect((await data.load()).messages).toMatchSnapshot()
  })

  it("should compile catalog with relative path with no warnings", async () => {
    const built = await build(
      path.join(__dirname, "./relative-catalog-path/entrypoint.js")
    )

    expect(built.stats.errors).toEqual([])
    expect(built.stats.warnings).toEqual([])

    const data = await built.loadBundle()
    expect((await data.load()).messages).toMatchSnapshot()
  })

  it("should throw an error when requested catalog don't belong to lingui config", async () => {
    const built = await build(
      path.join(__dirname, "./not-known-catalog/entrypoint.js")
    )

    expect(built.stats.errors[0].message).toContain(
      "is not matched to any of your catalogs paths"
    )
    expect(built.stats.warnings).toEqual([])
  })

  it("should trigger webpack recompile on catalog dependency change", async () => {
    const fixtureTempPath = await copyFixture(path.join(__dirname, "po-format"))

    const watching = watch(path.join(fixtureTempPath, "/entrypoint.js"))

    const res = await watching.build()

    expect((await res.loadBundle().then((m) => m.load())).messages)
      .toMatchInlineSnapshot(`
      {
        ED2Xk0: String from template,
        mVmaLu: [
          My name is ,
          [
            name,
          ],
        ],
        mY42CM: Hello World,
      }
    `)

    // change the dependency
    await fs.writeFile(
      path.join(fixtureTempPath, "/locale/messages.pot"),
      `msgid "Hello World"
msgstr ""

msgid "My name is {name}"
msgstr ""

msgid "String from template changes!"
msgstr ""
`
    )

    const stats2 = await watching.build()
    jest.resetModules()

    expect((await stats2.loadBundle().then((m) => m.load())).messages)
      .toMatchInlineSnapshot(`
      {
        mVmaLu: [
          My name is ,
          [
            name,
          ],
        ],
        mY42CM: Hello World,
        wg2uwk: String from template changes!,
      }
    `)

    await watching.stop()
  })
})

async function copyFixture(srcPath: string) {
  const fixtureTempPath = mkdtempSync(
    path.join(os.tmpdir(), `lingui-test-fixture-${process.pid}`)
  )

  await fs.cp(srcPath, fixtureTempPath, {
    recursive: true,
  })

  return fixtureTempPath
}
