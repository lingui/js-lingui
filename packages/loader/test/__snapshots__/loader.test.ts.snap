// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lingui-loader should compile catalog in json format 1`] = `
{
  key: Message,
  key2: [
    Hello ,
    [
      name,
    ],
  ],
}
`;

exports[`lingui-loader should compile catalog in po format 1`] = `
{
  ED2Xk0: String from template,
  mVmaLu: [
    My name is ,
    [
      name,
    ],
  ],
  mY42CM: Hello World,
}
`;

exports[`lingui-loader should compile catalog with relative path with no warnings 1`] = `
{
  ED2Xk0: String from template,
  mVmaLu: [
    My name is ,
    [
      name,
    ],
  ],
  mY42CM: Hello World,
}
`;
